examples/wasm/wasm_sample/Makefile

```diff
@@ -0,0 +1,101 @@
+LLVM_VERSION = 11
+
+#LINK_FLAGS_bak =\
+#		-nostartfiles \
+#		--nostdlib \
+#		--nostdinc \
+#		--print-map\
+#		--initial-memory= \
+#		--allow-undefined-file=<value>
+#		--initial-memory=65536
+
+# --export-all           Export all symbols (normally combined with --no-gc-sections)
+# --export-dynamic       Put symbols in the dynamic symbol table
+# --export-table         Export function table to the environment
+# --export=<value>       Force a symbol to be exported
+# --fatal-warnings       Treat warnings as errors
+# --import-memory        Import memory from the environment
+# --import-table         Import function table from the environment
+# --initial-memory=<value>
+#                        Initial size of the linear memory
+# --lto-O<opt-level>     Optimization level for LTO
+# --lto-partitions=<value>
+#                        Number of LTO codegen partitions
+# -L <dir>               Add a directory to the library search path
+# -l <libName>           Root name of library to use
+# --max-memory=<value>   Maximum size of the linear memory
+# --merge-data-segments  Enable merging data segments
+# --mllvm <value>        Options to pass to LLVM
+# --pie                  Create a position independent executable
+# --print-gc-sections    List removed unused sections
+# --relocatable          Create relocatable object file
+# --shared-memory        Use shared linear memory
+# --shared               Build a shared object
+# --stack-first          Place stack at start of linear memory rather than after data
+# --strip-all            Strip all symbols
+# --strip-debug          Strip debugging information
+# -S                     Alias for --strip-debug
+# -s                     Alias for --strip-all
+# --thinlto-cache-dir=<value>
+#                        Path to ThinLTO cached object file directory
+# --thinlto-cache-policy=<value>
+#                        Pruning policy for the ThinLTO cache
+# --thinlto-jobs=<value> Number of ThinLTO jobs
+# --threads              Run the linker multi-threaded
+# --undefined=<value>    Force undefined symbol during linking
+# --verbose              Verbose mode
+# --version              Display the version number and exit
+# -z <option>            Linker option extensions
+
+
+LINK_FLAGS = --export main \
+		--allow-undefined \
+		--export-dynamic \
+		--no-entry \
+		--strip-all \
+		-error-limit=0 \
+		--lto-O3 \
+		-O3 \
+		--gc-sections\
+		-z stack-size=4096 \
+
+#COMPILE_FLAGS_mov=\
+		-Wl,--allow-undefined\
+		-Wl,--strip-debug \
+		-nostartfiles \
+		-std=c++14 \
+
+COMPILE_FLAGS = -Wall \
+		--target=wasm32-unknown-unknown-wasm \
+		-emit-llvm \
+		-Os \
+		-flto \
+		-fvisibility=hidden \
+		-ffunction-sections \
+		-fdata-sections \
+
+%.show: %.wasm
+	wasm2wat $<
+
+%.wasm: %.o Makefile
+	wasm-ld-$(LLVM_VERSION) -o $@ $(LINK_FLAGS) $<
+
+
+%.o: %.cpp Makefile FORCE
+	clang++-$(LLVM_VERSION) \
```

Comment:
the `LLVM_VERSION` is used for the wasm tools `wasm-ld-<version>` and co at least for ubuntu-focal they are not liked to `wasm-ld` the only reason they are also with the clang is that i don't want to have 2 tool-chains being used for one build.

Reply:
> at least for ubuntu-focal they are not liked to `wasm-ld`

:frowning_face: [Arch Linux](https://archlinux.org/packages/extra/x86_64/lld/), [Fedora](https://koji.fedoraproject.org/koji/rpminfo?rpmID=27479493), [Alpine Linux](https://pkgs.alpinelinux.org/contents?branch=edge&name=lld&arch=x86_64&repo=community) and likely all distros except Ubuntu does it different: There is only `wasm-ld` but no `wasm-ld-<VERSION>`.

> i don't want to have 2 tool-chains being used for one build.

Note that e.g. GCC doesn't have its own linker. So far, I never had problems with updating the linker or even wildly mixing linker and compiler (e.g. using gold with gcc, using binutil's ld with gcc, using lld with gcc, using gold with clang, ...).

I think the reason there is no lld-<version> (or wasm-ld-<version>) is that there is no advantage in having multiple versions of it installed. For compilers new optimization features regularly break existing code (if relies on undefined behavior or behavior that only since recently is considered undefined), so one might want an older version to still get working binaries until the source is fixed.

So the question now is: Should this work by default for Ubuntu, or for everyone else. I'm all for the latter, but I'm not an Ubuntu user. (Honestly, isn't Ubuntu the Windows among the Linux Distros?)

Reply:
Just checked Debian and sure it has a wasm-ld (WTF is wrong with ~(my)~ Ubuntu). 

     (I reinstalled the lld package and got still no wasm-ld link -> not a incomplete pkg install) -> not my

Ubuntu has  `wasm-ld-11 wasm-ld-8` and  `lld-8, lld-11, lld`

I will change it to default wasm-ld and add a fallback for Ubuntu.

The RIOT-Docker uses Ubuntu

Why did i put Kubuntu on this Laptop  (I like Debian and they said it is like Debian and better (userfriendly) but no it is worse and i think it is also very unfriendly to me)- somehow i think i should try if `apt install debian` would work

Reply:
just check my fallback thing result wasm-ld-8 does not like clang (11) object files

Reply:
> just check my fallback thing result wasm-ld-8 does not like clang (11) object files

Maybe this is just a bug / missing feature of the old lld? I wouldn't bother too much if it works the other way round (recent lld, legacy clang). To me it seems like the package of `lld` is up to date (and so is the package of `clang`), while there are additional `clang-<ver>` packages if one needs older compilers to compile non-standard-conformant code.

Reply:
My tests showed that mixing lld (wasm-ld) and clang versions does not work -> try to avoid that.